### PR TITLE
Job progressions

### PR DESCRIPTION
**Pull request by Fausto95**
Adding a job to migrate all users' bad progressions' answers 

- [x] I ensured my code is tested:

### CODE REVIEW

File: app.json [hunk 1]
```diff
@@ -2,5 +2,6 @@
   "name": "Coorpacademy",
   "displayName": "Coorpacademy",
   "assistanceEmail": "assistance@coorpacademy.com",
+  "asyncStorageVersion": "1",
```
Comment (Fausto95) [inline app.json]:
```suggestion
```

File: src/screens/authentication.js [hunk 2]
```diff
@@ -98,6 +98,16 @@ class AuthenticationScreen extends React.PureComponent<Props, State> {
     this.handleDetailsNavigation(AUTHENTICATION_TYPE.MAGIC_LINK);
   };
 
+  runMigrations = async () => {
+    try {
+      await migrationsRunner();
+      /* istanbul ignore next */
```
Comment (Fausto95) [inline src/screens/authentication.js]: don't blame me 😆

File: src/migrations/index.js [hunk 3]
```diff
@@ -1,45 +1,11 @@
...
-    _.filter(([version]) => version > currentASVersion),
-    _.sortBy('0'),
-    vm =>
-      Promise.mapSeries(vm, async ([version, [predicate, migration]]) => {
-        await applyMigration(predicate, migration)(asyncStorage);
-        await asyncStorage.setItem(asyncStorageVersionKey, version);
-      })
-  )(migrations);
+const migrations: Migrations = {
+  '2': [fixProgressionsAnswers.predicate, fixProgressionsAnswers.transformer]
```
Comment (VincentCATILLON) [inline src/migrations/index.js]: why 2 ?

File: src/migrations/scripts/fix-progressions-answers/tests/fixtures.json [hunk 4]
```diff
@@ -0,0 +1,137 @@
+{
```
Comment (VincentCATILLON) [inline src/migrations/scripts/fix-progressions-answers/tests/fixtures.json]: no, typed JS please

File: src/screens/authentication.js (hunk 2 above)
Comment (VincentCATILLON) [reply src/screens/authentication.js]: no, no and no

File: src/screens/authentication.js [hunk 5]
```diff
@@ -98,6 +98,16 @@ class AuthenticationScreen extends React.PureComponent<Props, State> {
...
+      return this.hideSplashScreen();
+    } catch (error) {
+      return this.hideSplashScreen();
```
Comment (VincentCATILLON) [inline src/screens/authentication.js]: why do you hide the splash screen twice ? do it after the try catch

File: src/migrations/run-migrations.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
+// @flow
```
Comment (VincentCATILLON) [inline src/migrations/run-migrations.js]: why this file is need as the migration system is already merged into master ?

Comment (Fausto95) [reply src/migrations/run-migrations.js]: I changed the name

File: src/screens/authentication.js (hunk 5 above)
Comment (Fausto95) [reply src/screens/authentication.js]: I don't hide it twice, and hide when the job's promise is fullfilled or rejected

File: src/migrations/index.js (hunk 3 above)
Comment (Fausto95) [reply src/migrations/index.js]: ?

File: src/migrations/run-migrations.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      if (value !== nextValue) {
+        await asyncStorage.setItem(key, nextValue);
+      }
+    }
+  });
+};
+
+const applyMigrations = (asyncStorageVersionKey, migrations) => async (
+  asyncStorage: typeof AsyncStorage
+) => {
```
Comment (VincentCATILLON) [inline src/migrations/run-migrations.js]: return type

File: src/migrations/run-migrations.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
+// @flow
+import Promise from 'bluebird';
+import AsyncStorage from '@react-native-community/async-storage';
+import _ from 'lodash/fp';
+import type {Migrations} from './types';
+
+const applyMigration = (predicate, migration) => async (asyncStorage: typeof AsyncStorage) => {
```
Comment (VincentCATILLON) [inline src/migrations/run-migrations.js]: return type

File: src/migrations/index.js (hunk 3 above)
Comment (Fausto95) [reply src/migrations/index.js]: currently the AS version is not defined, then we assume that is v1, so this script will bump it to v2

File: src/migrations/types.js [hunk 9]
```diff
@@ -1,7 +1,7 @@
 // @flow
 
 export type Predicate = (key: string) => boolean;
-export type Transformer = (value: string, key: string) => Promise<string>;
+export type Transformer = (value: string, key: string) => Promise<string> | string;
```
Comment (VincentCATILLON) [inline src/migrations/types.js]: Can you explain ?

File: src/migrations/scripts/fix-progressions-answers/tests/fixtures.json (hunk 4 above)
Comment (Fausto95) [reply src/migrations/scripts/fix-progressions-answers/tests/fixtures.json]: Needless

File: src/migrations/types.js (hunk 9 above)
Comment (Fausto95) [reply src/migrations/types.js]: The transformer can return a promise<string> or a string(simple transform case).

File: src/redux/middlewares/progressions-synchronization.js [hunk 10]
```diff
@@ -20,7 +20,7 @@ const createMiddleware = ({
 Middleware<State, StoreAction<Action>, Dispatch<StoreAction<Action>>> => ({
   dispatch,
   getState
-}) => next => action => {
+}) => next => async action => {
```
Comment (VincentCATILLON) [inline src/redux/middlewares/progressions-synchronization.js]: no return type ?

File: src/screens/authentication.js (hunk 2 above)
Comment (Fausto95) [reply src/screens/authentication.js]: I don't know what I shall do since the script is tested already

File: src/migrations/types.js (hunk 9 above)
Comment (Volubyl) [reply src/migrations/types.js]: this return type seems highly akward. How could I know when this function in synchrone or not ? I believe you should decided if the function should be sync or not.

File: src/screens/authentication.js (hunk 5 above)
Comment (Volubyl) [reply src/screens/authentication.js]: like Vincent is there any difficulties to put `return this.hideSplashScreen();`  after the try catch ?

Comment (Volubyl) [reply src/screens/authentication.js]: in the `catch` you could throw a simple error and handle it eslewhere

File: src/migrations/run-migrations.js (hunk 6 above)
Comment (VincentCATILLON) [reply src/migrations/run-migrations.js]: Ok i didn't know, imo `runner` should be enough

File: android/app/build.gradle [hunk 11]
```diff
@@ -55,7 +55,7 @@ android {
         multiDexEnabled true
         missingDimensionStrategy 'react-native-camera', 'general'
         versionCode 1
-        versionName "2.0.39"
+        versionName "2.0.40"
```
Comment (VincentCATILLON) [inline android/app/build.gradle]: should not be there

File: ios/Coorpacademy-tvOS/Info.plist [hunk 12]
```diff
@@ -15,7 +15,7 @@
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>2.0.39</string>
+	<string>2.0.40</string>
```
Comment (VincentCATILLON) [inline ios/Coorpacademy-tvOS/Info.plist]: should not be there

File: ios/Coorpacademy-tvOSTests/Info.plist [hunk 13]
```diff
@@ -15,7 +15,7 @@
 	<key>CFBundlePackageType</key>
 	<string>BNDL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>2.0.39</string>
+	<string>2.0.40</string>
```
Comment (VincentCATILLON) [inline ios/Coorpacademy-tvOSTests/Info.plist]: should not be there

File: ios/Coorpacademy/Info.plist [hunk 14]
```diff
@@ -26,7 +26,7 @@
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>2.0.39</string>
+	<string>2.0.40</string>
```
Comment (VincentCATILLON) [inline ios/Coorpacademy/Info.plist]: should not be there

File: ios/CoorpacademyTests/Info.plist [hunk 15]
```diff
@@ -15,7 +15,7 @@
 	<key>CFBundlePackageType</key>
 	<string>BNDL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>2.0.39</string>
+	<string>2.0.40</string>
```
Comment (VincentCATILLON) [inline ios/CoorpacademyTests/Info.plist]: should not be there

File: package.json [hunk 16]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "coorpacademy-mobile",
-  "version": "2.0.39",
+  "version": "2.0.40",
```
Comment (VincentCATILLON) [inline package.json]: should not be there

Comment (Fausto95) [inline package.json]:
```suggestion
  "version": "2.0.39",
```

File: ios/CoorpacademyTests/Info.plist (hunk 15 above)
Comment (Fausto95) [inline ios/CoorpacademyTests/Info.plist]:
```suggestion
	<string>2.0.39</string>
```

File: ios/Coorpacademy/Info.plist (hunk 14 above)
Comment (Fausto95) [inline ios/Coorpacademy/Info.plist]:
```suggestion
	<string>2.0.39</string>
```

File: ios/Coorpacademy-tvOSTests/Info.plist (hunk 13 above)
Comment (Fausto95) [inline ios/Coorpacademy-tvOSTests/Info.plist]:
```suggestion
	<string>2.0.39</string>
```

File: ios/Coorpacademy-tvOS/Info.plist (hunk 12 above)
Comment (Fausto95) [inline ios/Coorpacademy-tvOS/Info.plist]:
```suggestion
	<string>2.0.39</string>
```

File: android/app/build.gradle (hunk 11 above)
Comment (Fausto95) [inline android/app/build.gradle]:
```suggestion
        versionName "2.0.39"
```

**Comment by VincentCATILLON** [comment]
Comment: the flow coverage is really low in the migrations folder, because of lodash usage, it should be enhanced

**Pull request merged**